src/primitives/block.cpp

```diff
@@ -109,6 +112,47 @@ uint256 CBlock::CheckMerkleBranch(uint256 hash, const std::vector<uint256>& vMer
     return hash;
 }
 
+static uint64_t next_pow2(uint64_t x)
+{
+    x -= 1;
+    x |= (x >> 1);
+    x |= (x >> 2);
+    x |= (x >> 4);
+    x |= (x >> 8);
+    x |= (x >> 16);
+    x |= (x >> 32);
+    return x + 1;
+}
+
+uint256 CBlock::BuildAuthDataMerkleTree() const
+{
+    std::vector<uint256> tree;
+    auto perfectSize = next_pow2(vtx.size());
+    tree.reserve(perfectSize * 2); // Safe upper bound for the number of total nodes.
+
+    // Add the leaves to the tree. v1-v4 transactions will append empty leaves.
+    for (auto &tx : vtx) {
+        tree.push_back(tx.GetAuthDigest());
+    }
+    // Append empty leaves until we get a perfect tree.
+    tree.insert(tree.end(), perfectSize - vtx.size(), uint256());
+
+    int j = 0;
+    for (int layerWidth = tree.size(); layerWidth > 1; layerWidth = (layerWidth + 1) / 2) {
```

Comment:
Why `layerWidth + 1`, rather than `layerWidth`? (It is a loop invariant that `layerWidth` is a power of 2 greater than 1.)

Reply:
This line was copied from the txid Merkle tree impl, which doesn't have this guarantee. I can simplify this.